File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/Movement.java
package org.firstinspires.ftc.teamcode.utils;

import com.pedropathing.follower.Follower;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

public abstract class Movement extends LinearOpMode {
    public abstract void systemInit();
    public void systemStart() {

    }
    public abstract void systemLoop();

    @Override
    public void runOpMode() {
        systemInit();
        Follower follower = new Follower(hardwareMap);

        waitForStart();
        follower.startTeleopDrive();
        systemStart();

        while (opModeIsActive()) {
            follower.setTeleOpMovementVectors(gamepad1.left_stick_x, gamepad1.left_stick_y, gamepad1.right_stick_x, false);
            follower.update();

            systemLoop();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/utils/systems/Movement.kt
package org.firstinspires.ftc.utils.systems

import com.pedropathing.follower.Follower
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode

abstract class Movement : LinearOpMode() {
    /**
     * Initializes the system or subsystem along with
     * any hardware devices not related to the movement of the robot.
     */
    abstract fun systemInit()
    /**
     * Actions to be done after the start button is pressed but only once.
     * This function is empty by default.
     */
    open fun systemStart() {}
    /**
     * Actions to be done after the stop button is pressed.
     */
    abstract fun systemLoop()
    /**
     * The main function that runs the system or subsystem.
     */
    override fun runOpMode() {
        /**
         * Pedro Pathing`s follower
         */
        val follower = Follower(hardwareMap)
        // Initialize the system or subsystem
        systemInit()
        // Wait for the start button to be pressed
        waitForStart()
        // Initialize the follower for teleop drive
        follower.startTeleopDrive()
        // Does all of the actions specified in the start button
        systemStart()
        // Run the op mode
        while (opModeIsActive()) {
            // Sets the movement vectors for the follower
            follower.setTeleOpMovementVectors(
                gamepad1.left_stick_x.toDouble(),
                gamepad1.left_stick_y.toDouble(),
                gamepad1.right_stick_x.toDouble(),
                false)
            // Updates the follower
            follower.update()
            // Run the system or subsystem
            systemLoop()
        }
    }
}